fix(mi): Update applet permissions (M2-6721, M2-6727, M2-6728)

ChaconC opened the pull request:
- [x] Tests for the changes have been added
- [ ] Delivered the `fix` or `feature` branches into `develop` or `release` branches via `Squash and Merge` (to keep clean history)

### 📝 Description

Adds additional permission checks for removing certain elements for `Coordinator`, `Reviewer` and `Editor` roles

🔗 [Jira Ticket M2-6721](https://mindlogger.atlassian.net/browse/M2-6721)
🔗 [Jira Ticket M2-6727](https://mindlogger.atlassian.net/browse/M2-6727)
🔗 [Jira Ticket M2-6728](https://mindlogger.atlassian.net/browse/M2-6728)

Changes include:

- Add `checkIfCanAccessData` util
- Remove "Add Activity" buttons for Reviewers, Coordinator roles
- Remove "Add Participant" buttons for Reviewer roles
- Remove navigation to DataViz for Coordinator, Editor roles
- Remove data export buttons for Coordinator, Editor roles

### 📸 Screenshots

<!--
If your work here contains visual changes, provide before (optional) and after screenshots, GIFs, or videos.

If not, then delete this section
-->

### Applet > Activities (Coordinator)
* No Settings button
* No Add Activity button
* No Export buttons

| Before                      | After                                 |
| -------------------------------------- | ------------------------------------- |
| ![CleanShot 2024-05-28 at 16 29 30](https://github.com/ChildMindInstitute/mindlogger-admin/assets/2916156/eaec8cc3-db58-4ad8-b7ce-3e3e15a98efb) | ![CleanShot 2024-05-28 at 16 30 12](https://github.com/ChildMindInstitute/mindlogger-admin/assets/2916156/30d76c77-c80e-4bef-a558-bc6ca663c024) |

### Applet > Activities (Reviewer)
* No Settings button
* No Add Activity button

| Before                      | After                                 |
| -------------------------------------- | ------------------------------------- |
| ![CleanShot 2024-05-29 at 08 21 55](https://github.com/ChildMindInstitute/mindlogger-admin/assets/2916156/bffff236-a321-4e7d-877a-90180b99e6b4) | ![CleanShot 2024-05-29 at 08 21 41](https://github.com/ChildMindInstitute/mindlogger-admin/assets/2916156/963962cc-5b56-4bea-ab48-ccbd9743e758) |

### Applet > Participants (Reviewer)
* No Add Participant button
* No Settings button

| Before                      | After                                 |
| -------------------------------------- | ------------------------------------- |
| ![CleanShot 2024-05-29 at 08 25 26](https://github.com/ChildMindInstitute/mindlogger-admin/assets/2916156/b3d8fec5-6375-4bc0-9364-a2682a6f13df) | ![CleanShot 2024-05-29 at 08 25 14](https://github.com/ChildMindInstitute/mindlogger-admin/assets/2916156/d3e66441-9079-4827-8a77-f47897168703) |


### Hiding Dots Menu 
| Before (Optional)                      | After                                 |
| -------------------------------------- | ------------------------------------- |
| ![CleanShot 2024-05-28 at 16 33 11](https://github.com/ChildMindInstitute/mindlogger-admin/assets/2916156/0eca087f-dc9a-41bc-8044-73d953346634) | ![CleanShot 2024-05-28 at 16 33 03](https://github.com/ChildMindInstitute/mindlogger-admin/assets/2916156/c02ef02b-ecd2-4f01-a117-d2560990077a) |

### 🪤 Peer Testing

#### Setup

* Create an Applet with Reviewer, Coordinator and Editor accounts
* The applet should have at least 1 response so that DataViz navigation works correctly

#### Testing
**Log in as Reviewer**
* Select shared workspace
* Navigate to Applet > Activities
  * Validate no "Add Activity" button 
* Navigate to Applet > Participants
  * Validate no "Add Participant" button
  * Validate no Settings button
* Select a Participant
* Navigate to DataViz

**Log in as Coordinator**
* Select shared workspace
* Navigate to Applet > Activities
  * Validate no "Add Activity" button 
* Navigate to Applet > Participants
  * Validate no Export button
  * Validate no Settings button
* Select a Participant
* Confirm no navigation to DataViz

### ✏️ Notes

Expected permissions can be reviewed in the [Confluence Page](https://mindlogger.atlassian.net/wiki/x/FABSCg) checklists for each role.

| Role | View Participants | Add/Edit Participants | Add/Edit Activity | Data Export/Visualization | Take Now | Assign* |
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
| Owner/Manager | ✅ | ✅ | ✅ | ✅ | ✅ |  ✅ | 
| Coordinator | ✅ | ✅ | ❌ | ❌ | ❌ |  ✅ | 
| Reviewer | ✅ | ❌ | ❌ | ✅ | ✅ | ❌ | 
| Editor* | ❌ | ❌ | ✅ | ❌ | ❌ | ❌ | 

* Assign button is controlled by the `enableActivityAssign` feature flag

**Note about Editor Role:**
~Because Editors are not allowed to view participants, all routes inside an applet will fail to render due to being unable to retrieve `/respondents`. Will link a related ticket for this issue.~

The previously mentioned Editor Role limitation has been addressed and Editors should be able to view content in the Activities tab only.

ChaconC committed:
Add checkIfCanAccessData util

ChaconC committed:
Check permissions for showing Export menu item

ChaconC committed:
Hide ActionsMenu button if there are no items

ChaconC committed:
Check permission for ActivitySummaryCard nav

ChaconC committed:
Check permissions for showing Export menu item

ChaconC committed:
update tests

ChaconC committed:
Add canAccessData to Participants table

ChaconC committed:
fix checkIfCanAccessData test

ChaconC committed:
pass roles to getParticipantActions test

farmerpaul commented on src/modules/Dashboard/features/Participants/Participants.utils.tsx:
```diff
@@ -31,12 +32,16 @@ export const getParticipantActions = ({
   const isPending = status === RespondentStatus.Pending;
   const isEditable = !!filteredApplets?.editable.length;
   const isViewable = !!filteredApplets?.viewable.length;
+  const canEdit = checkIfCanEdit(roles) && isViewable && !isPending;
```
I'm a little confused by the difference between `isEditable` and `canEdit`… they are defined slightly differently. Could we get away with using just one, or if not, name them more distinctly?

ChaconC replied:
Good catch, looking at the file history this was a remnant from my original approach, then I added `checkIfCanAccessData` and replaced `canEdit` with `canAccessData` but didn't remove this boolean.

I'll move the role validation to `isEditable` and we can get rid of `canEdit`

sultanofcardio commented on src/modules/Dashboard/features/Participants/Participants.tsx:
```diff
@@ -418,7 +422,7 @@ export const Participants = () => {
             data-testid={`${dataTestid}-search`}
           />
-          {appletId && (
+          {canAddParticipant && (
```
Remember to add this to the empty state add participant button below as well

<img src="https://github.com/ChildMindInstitute/mindlogger-admin/assets/14842108/1e209800-b7a0-43c5-87b5-8bfe00e2fb37" width="480">

farmerpaul commented on src/modules/Dashboard/components/FlowGrid/FlowGrid.hooks.tsx:
```diff
@@ -31,17 +31,15 @@ export function useFlowGridMenu({
+    (isManagerOrOwner(roles?.[0]) || roles?.includes(Roles.SuperAdmin));
+
+  const canAccessData = checkIfCanAccessData(roles);
+  const showDivider = (canEdit || canAccessData) && canDoTakeNow;
```
```suggestion
  const showDivider = (canEdit || canAccessData) && (featureFlags.enableActivityAssign || canDoTakeNow);
```

JustinNusca commented on src/modules/Dashboard/components/FlowGrid/FlowGrid.hooks.tsx:
```diff
@@ -31,17 +31,15 @@ export function useFlowGridMenu({
-      roles?.includes(Roles.SuperAdmin));
+    (isManagerOrOwner(roles?.[0]) || roles?.includes(Roles.SuperAdmin));
+
+  const canAccessData = checkIfCanAccessData(roles);
```
👍 When you messaged me the other day about this, I was already thinking "why aren't we just using a shared util fn here" — I wasn't even aware of `checkIfCanEdit` previously.

JustinNusca commented on src/modules/Dashboard/components/FlowGrid/FlowGrid.hooks.tsx:
```diff
@@ -31,17 +31,15 @@ export function useFlowGridMenu({
-    (isManagerOrOwner(roles?.[0]) ||
-      roles?.includes(Roles.Coordinator) ||
-      roles?.includes(Roles.SuperAdmin));
+    (isManagerOrOwner(roles?.[0]) || roles?.includes(Roles.SuperAdmin));
```
Any particular reason you opted not to abstract this to a similar  "checkRoles" utility fn? I notice this is the same check made in `ActivityGrid.utils.tsx`.

ChaconC replied:
No reason really, was just limiting my changes. I'll add it to `checkRoles` and add test cases 👍

JustinNusca commented on src/modules/Dashboard/features/Applet/Activities/ActivitiesToolbar/ActivitiesToolbar.tsx:
```diff
@@ -19,6 +21,10 @@ export const ActivitiesToolbar = ({
   const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
+  const workspaceRoles = workspaces.useRolesData();
+  const roles = appletId ? workspaceRoles?.data?.[appletId] : undefined;
+
```
Extremely nit-picky, and I'm not opposed to linebreaks for readability or whatever, but is the extra line break here intentional? 🧐

sultanofcardio commented on src/modules/Dashboard/features/Applet/Activities/ActivitiesToolbar/ActivitiesToolbar.tsx:
```diff
@@ -74,15 +80,17 @@ export const ActivitiesToolbar = ({
+                sx={{ minWidth: theme.spacing(13.2) }}
+              >
+                {t('addActivity')}
+              </Button>
```
In addition to this, according to scenario 3 of [M2-6727](https://mindlogger.atlassian.net/browse/M2-6727), it sounds like the reviewer should not be able to see the activities tab either. I'm not sure if that's what was meant, since the language was vague but I thought I'd point that out. Are you clear on this?

In case it turns out they shouldn't, I'm thinking of what the solution would be. In our multi-informant environment we use a pop-up like this to block access to the team tab

<img width="1327" alt="image" src="https://github.com/ChildMindInstitute/mindlogger-admin/assets/14842108/1b4d8fa6-3b57-4185-a4e4-1e5ac1b7a1c4">

However, for the schedule tab I see something like this, which feels a little cleaner

<img width="1323" alt="image" src="https://github.com/ChildMindInstitute/mindlogger-admin/assets/14842108/a03a9729-eef6-41ac-b9f5-ed6c24f5efe2">

JustinNusca replied:
I wanted to piggy-back off of this comment, because I ran into a couple similar issues. 

I'm not sure if it's within the scope of these tickets, which seem very specific about particular buttons/menu items, rather than access to tabs per-se, but I think it's worth discussing.

While the Applet-level Schedule tab has a nice "Forbidden" empty state for Reviewers, the Schedule tab at the Participant level raises the same "No access to applet" alert and redirects to the applets list:
 
![Screenshot 2024-05-29 at 4 57 00 PM](https://github.com/ChildMindInstitute/mindlogger-admin/assets/1670836/9e67c94c-69ea-483e-a3aa-c2d5be72ce99)

It seems like maybe we should be consistent in our behaviour here? I think the "Forbidden" empty states are preferable and less disruptive, but if that's the case then we need updated empty states for the Participants and Teams tab as well — the default empty states for both include buttons prompting the user to "Add a participant/manager", which ought to be conditional based on role. (eg, per @sultanofcardio's [other comment](https://github.com/ChildMindInstitute/mindlogger-admin/pull/1761#discussion_r1619563399).

ChaconC replied:
@sultanofcardio went ahead and added permission checks to return the `NoPermissions` component instead of the modal. This should fix the problem you highlighted for Reviewer and Coordinator roles, and also applied the same changes for all Editor tabs

@JustinNusca This wasn't a huge lift so I've added the same behavior in the Participant > Schedule for Reviewers and Editors

farmerpaul commented on src/modules/Dashboard/components/ActivityGrid/ActivityGrid.utils.tsx:
```diff
@@ -44,7 +42,7 @@ export const getActivityActions = ({
       title: t('exportData'),
       context: { appletId, activityId },
-      isDisplayed: true,
+      isDisplayed: canAccessData,
```
In `FlowGrid.hooks.tsx`, the Export Data menu item has `disabled: true` at all times, not yet sure why. Presumably because it's unimplemented? However, here, it's not disabled, but when it's available (i.e. when `canAccessData == true`), it just shows a TODO alert.

Do you know if this action is expected to be implemented by launch? In any case, it seems that `ActivityGrid` and `FlowGrid` should treat this menu item the same…

ChaconC replied:
Not sure why `disabled: true` was added but also thought it was odd, maybe @JustinNusca can chime in.

Another thing, yes this functionality is meant for R1 and removing the TODO alerts will be covered by [M2-6736](https://mindlogger.atlassian.net/browse/M2-6736)

JustinNusca replied:
I meant to reply earlier to speak to this — yeah, I set `disabled: true` initially.

At the time, it seemed like we were in the practice of adding a lot of placeholder UI for functionality that wasn't yet present. I opted to just set `disabled`: it felt less confusing/disruptive from the user perspective than having an enabled-but-nonfunctional button and/or alert prompts, with the expectation that the presence of the menu item would be clear enough for whoever eventually needed to wire it up.

In the end, it seems like all these "Export Data" options had the requisite functionality for supporting these options all along, and yeah, there's a couple tasks currently scheduled for Sprint 10 that should handle them.

sultanofcardio commented on src/shared/utils/checkRole.ts:
```diff
@@ -11,3 +11,10 @@ export const checkIfCanEdit = (roles?: Roles[]) =>
+  Boolean(
+    isManagerOrOwner(roles?.[0]) ||
+      roles?.includes(Roles.SuperAdmin) ||
+      roles?.includes(Roles.Reviewer),
```
There seems to be conflicting information in these tickets. Scenario 4 of [M2-6727](https://mindlogger.atlassian.net/browse/M2-6727) says that “A reviewer can only view or export data.” However, the 3rd case of [M2-6728](https://mindlogger.atlassian.net/browse/M2-6728) says that “Reviewer should not be able to export data in shared account”.

Based on [Cari's comment](https://mindlogger.atlassian.net/browse/M2-6727?focusedCommentId=21467), it seems the reviewer should be able to export data so maybe we should get the description of [M2-6728](https://mindlogger.atlassian.net/browse/M2-6728) updated

mbanting replied:
To potentially further narrow the permissions for a reviewer, another [comment](https://mindlogger.atlassian.net/browse/M2-6727?focusedCommentId=21467) by Cari and an additional reference in [Confluence](https://mindlogger.atlassian.net/wiki/spaces/MINDLOGGER1/pages/172589139/Invite+Reviewer+Checklist) states Reviewer can 
> view or export ONLY assigned user's (Respondents) data

ChaconC replied:
@mbanting correct, that's the comment I went with, and Reviewers only see the participants they have assigned + team members so this seems to cover the criteria.

Edit: Added a note for QA in M2-6728 for clarification, but my take is Reviewers are able to export data for their assigned participants.

sultanofcardio replied:
It's the same comment, just wanted to highlight this so we could clarify for QA testing. I'll resolve this thread now 👍

JustinNusca commented on src/modules/Dashboard/features/Participant/Activities/Activities.tsx:
```diff
@@ -128,6 +134,12 @@ export const Activities = () => {
     );
   };
 
+  const getClickHandler = () => {
+    if (!subjectId || !appletId || !canAccessData) return undefined;
+
+    return onClickItemHandler;
+  };
```
If looks like the actual handler itself has an early return statement that makes the 2/3 of the same checks.

I wonder if it would be more conventional to just update [those lines](https://github.com/ChildMindInstitute/mindlogger-admin/pull/1761/files#diff-55e1e9f202a8f57998166fcac43a4bb07de7a8d708ad50847b1675c7c48033f8R114-R115) to be something like…

```jsx
const handleClickItem = (activityId: string) => {
  if (!subjectId || !appletId || !canAccessData) return;
  // …rest of fn…
```

… compared to adding a new wrapper function that conditionally returns a handler on the basis of mostly duplicate checks.

ChaconC replied:
I'm going to remove the check on the handler instead, I explicitly wanted to return `undefined` for those cases so that the cursor would change back to a pointer in the `ActivityCard` component (e.g. by having `onClick` be `undefined` as opposed to `() => void`)

sultanofcardio commented:
The tickets look like they need some story points as well

mbanting commented on src/modules/Dashboard/features/Applet/Activities/Activities.test.tsx:
```diff
@@ -166,12 +166,9 @@ describe('Dashboard > Applet > Activities screen', () => {
-      if (canEdit) {
+      const actionDots = screen.queryAllByTestId(`${testId}-activity-actions-dots`)[0];
+      if (actionDots && canEdit) {
+        userEvent.click(screen.getAllByTestId(`${testId}-activity-actions-dots`)[0]);
```
Can you use the existing reference to `actionDots` since its existence is being checked?

```suggestion
        userEvent.click(actionDots);
```

mbanting commented on src/modules/Dashboard/features/Applet/Activities/Activities.test.tsx:
```diff
@@ -211,12 +208,10 @@ describe('Dashboard > Applet > Activities screen', () => {
 
-        if (canDoTakeNow) {
+        if (actionDots && canDoTakeNow) {
+          userEvent.click(screen.getAllByTestId(`${testId}-activity-actions-dots`)[0]);
```
```suggestion
          userEvent.click(actionDots);
```

mbanting commented on src/modules/Dashboard/features/Applet/Activities/Activities.test.tsx:
```diff
@@ -259,12 +254,13 @@ describe('Dashboard > Applet > Activities screen', () => {
-        await waitFor(() => expect(screen.queryByTestId(`${testId}-activity-take-now`)).toBe(null));
+        const actionDots = screen.queryAllByTestId(`${testId}-activity-actions-dots`)[0];
+        if (actionDots) {
+          await userEvent.click(screen.getAllByTestId(`${testId}-activity-actions-dots`)[0]);
```
```suggestion
          await userEvent.click(actionDots);
```

mbanting commented on src/modules/Dashboard/features/Participant/Activities/Activities.test.tsx:
```diff
@@ -163,12 +163,9 @@ describe('Dashboard > Applet > Participant > Activities screen', () => {
-      if (canEdit) {
+      const actionDots = screen.queryAllByTestId(`${testId}-activity-actions-dots`)[0];
+      if (actionDots && canEdit) {
+        userEvent.click(screen.getAllByTestId(`${testId}-activity-actions-dots`)[0]);
```
```suggestion
        userEvent.click(actionDots);
```

mbanting commented on src/modules/Dashboard/features/Participant/Activities/Activities.test.tsx:
```diff
@@ -207,12 +204,9 @@ describe('Dashboard > Applet > Participant > Activities screen', () => {
-        if (canDoTakeNow) {
+        const actionDots = screen.queryAllByTestId(`${testId}-activity-actions-dots`)[0];
+        if (actionDots && canDoTakeNow) {
+          userEvent.click(screen.getAllByTestId(`${testId}-activity-actions-dots`)[0]);
```
```suggestion
          userEvent.click(actionDots);
```

mbanting commented on src/modules/Dashboard/features/Participant/Activities/Activities.test.tsx:
```diff
@@ -254,12 +248,17 @@ describe('Dashboard > Applet > Participant > Activities screen', () => {
-        await waitFor(() => expect(screen.queryByTestId(`${testId}-activity-take-now`)).toBe(null));
+        const actionDots = screen.queryAllByTestId(`${testId}-activity-actions-dots`)[0];
+        if (actionDots) {
+          await userEvent.click(screen.getAllByTestId(`${testId}-activity-actions-dots`)[0]);
```
```suggestion
          await userEvent.click(actionDots);
```

farmerpaul reviewed (commented):
Hey @ChaconC, tested and it seems to be working as described, though I haven't delved into the permissions as deeply as it looks like @sultanofcardio has (kudos to him!). I identified some other minor things in the code.

The other thing I identified is, when logged in as a Coordinator, navigating to **Applet > Participant Details > Schedule** tab shows an error alert, and kicks you out of the applet and workspace altogether. I think it should probably show the same thing that **Applet > Schedule** tab shows (a less jarring no-permissions state). Not sure if this is worth scoping into your current PR but it does seem related and hopefully a small thing?

mbanting commented on src/shared/components/ActionsMenu/ActionsMenu.tsx:
```diff
@@ -23,8 +23,18 @@ export const ActionsMenu = <T = unknown,>({
+    const displayedMenuItems = menuItems.filter((item) => 'isDisplayed' in item).length;
+    if (displayedMenuItems) {
+      return menuItems.filter((item) => item.isDisplayed === true).length > 0;
+    }
```
You can try using `some` instead of checking existence via the length

```suggestion
const hasMenuItems = useMemo(() => {
  // Check if there are any items with the 'isDisplayed' property
  const hasDisplayedProperty = menuItems.some((item) => 'isDisplayed' in item);

  // If there are items with 'isDisplayed' property, check if any of them are displayed
  if (hasDisplayedProperty) {
    return menuItems.some((item) => item.isDisplayed);
  }```

mbanting reviewed (dismissed):
Looks great @ChaconC . I tested the suggested peer testing and it's all working as specified.

Left a couple of minor suggestions.

ChaconC committed:
Apply suggestions from code review to tests

Co-authored-by: Marty <mbanting@yahoo.com>

ChaconC committed:
Use .some to validate displayed items instead of length

Co-authored-by: Marty <mbanting@yahoo.com>

ChaconC committed:
Update src/modules/Dashboard/components/FlowGrid/FlowGrid.hooks.tsx

Co-authored-by: Farmer Paul <paul.hh@metalab.com>

ChaconC committed:
minor cleanup

ChaconC committed:
remove redundant validation in clickHandler

ChaconC committed:
Refactor getParticipantActions and useFlowGridMenu validations

- Add checkIfCanManageParticipants, checkIfCanTakeNow to checkRoles

ChaconC committed:
Add role check to Add Participant button

ChaconC committed:
Add checkIfCanTakeNow checks to TakeNow buttons

ChaconC committed:
Add canAssignActivity checks

ChaconC committed:
Add TODOs to FlowGrid.hooks

ChaconC committed:
fix tests

ChaconC committed:
Add role checks for empty Tables

- Fixes Dashboard views for Editor role
- Prevent Reviewer from viewing Team content

ChaconC committed:
Fix Participant Schedule permissions for Reviewer

ChaconC committed:
Fix editor permissions

farmerpaul reviewed (commented):
Changes look good, thanks for making those! Before I approve, though, just wanted to point out one thing I'm experiencing:

I just tried testing access to an applet as a Reviewer (maybe that's what I did last time, too, even though I wrote Coordinator in my comment, can't remember), and I'm getting more of those applet access error popups when I click into Participant details. When it happens, I again get kicked out of both the applet and the workspace, which I presume shouldn't happen. I've recorded a video:

https://github.com/ChildMindInstitute/mindlogger-admin/assets/1282772/4d5f5474-3caf-47bb-ad29-f16d8ec93c77

mbanting reviewed (approved):
Thanks for addressing my suggestions/feedback!

ChaconC committed:
Remove checkIfCanTakeNow and update permissions

ChaconC committed:
remove unused canViewParticipants check

ChaconC commented:
> Changes look good, thanks for making those! Before I approve, though, just wanted to point out one thing I'm experiencing:
I just tried testing access to an applet as a Reviewer (maybe that's what I did last time, too, even though I wrote Coordinator in my comment, can't remember), and I'm getting more of those applet access error popups when I click into Participant details. When it happens, I again get kicked out of both the applet and the workspace, which I presume shouldn't happen. I've recorded a video:

@farmerpaul I've removed `Reviewer` from `TakeNow` permissions as this was an error on my part, and I updated the check  in `TakeNowModal` to prevent the error in the clip.

ChaconC committed:
Merge branch 'feature/multiinformant-metapod' into cc/M2-6721-applet-permissions

ChaconC committed:
update tests

JustinNusca reviewed (approved):
After your most recent changes, I'm no longer seeing the issue @farmerpaul described with the `reviewer` role, and the others appear to have the correct permissions reflected in the UI, too. 👍

Thanks for going ahead and updating the "forbidden"/empty states for the various screens, too!
